ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java

```diff
@@ -82,71 +82,126 @@ public void copyAndVerify(FileSystem destinationFs, Path destRoot,
           LOG.error("Failed to create destination directory: " + destination);
           throw new IOException("Destination directory creation failed");
         }
-        doCopyRetry(sourceFs, fileInfoList, destinationFs, destination, useRegularCopy);
 
-        // Verify checksum, retry if checksum changed
-        List<ReplChangeManager.FileInfo> retryFileInfoList = new ArrayList<>();
-        for (ReplChangeManager.FileInfo srcFile : srcFiles) {
-          if (!srcFile.isUseSourcePath()) {
-            // If already use cmpath, nothing we can do here, skip this file
-            continue;
-          }
-          String sourceChecksumString = srcFile.getCheckSum();
-          if (sourceChecksumString != null) {
-            String verifySourceChecksumString;
-            try {
-              verifySourceChecksumString
-                      = ReplChangeManager.checksumFor(srcFile.getSourcePath(), sourceFs);
-            } catch (IOException e) {
-              // Retry with CM path
-              verifySourceChecksumString = null;
-            }
-            if ((verifySourceChecksumString == null)
-                    || !sourceChecksumString.equals(verifySourceChecksumString)) {
-              // If checksum does not match, likely the file is changed/removed, copy again from cm
-              srcFile.setIsUseSourcePath(false);
-              retryFileInfoList.add(srcFile);
-            }
-          }
-        }
-        if (!retryFileInfoList.isEmpty()) {
-          doCopyRetry(sourceFs, retryFileInfoList, destinationFs, destination, useRegularCopy);
-        }
+		    // Copy files with retry logic on failure
+        doCopyRetry(sourceFs, fileInfoList, destinationFs, destination, useRegularCopy);
       }
     }
   }
 
-  private void doCopyRetry(FileSystem sourceFs, List<ReplChangeManager.FileInfo> fileList,
+  private void doCopyRetry(FileSystem sourceFs, List<ReplChangeManager.FileInfo> srcFileList,
                            FileSystem destinationFs, Path destination,
                            boolean useRegularCopy) throws IOException, LoginException {
     int repeat = 0;
-    List<Path> pathList = Lists.transform(fileList, ReplChangeManager.FileInfo::getEffectivePath);
+    boolean isCopyError = false;
+    List<Path> pathList = Lists.transform(srcFileList, ReplChangeManager.FileInfo::getEffectivePath);
```

Comment:
if we do transform here ..is it possible that some of the entries will be null ?
may be transform is not required here ?

Reply:
nope... here it won't return null. but just traverse through the list and get the effective path. So, cannot skip the transform.